source/detail/implementations/stylesheet.hpp

```diff
@@ -470,11 +470,21 @@ struct stylesheet
     format_impl *find_or_create_with(format_impl *pattern, const number_format &new_number_format, optional<bool> applied)
     {
         format_impl new_format = *pattern;
-        if (new_number_format.id() >= 164)
+        if (new_number_format.has_id() && new_number_format.id() < 164)
         {
-            find_or_add(number_formats, new_number_format);
+            new_format.number_format_id = new_number_format.id();
+        }
+        else
+        {
+            auto iter = std::find(number_formats.begin(), number_formats.end(), new_number_format);
+            if (iter == number_formats.end())
+            {
+                number_format new_item(new_number_format);
```

Comment:
Small optimisation: we can avoid the copy here. Just set the id after the `emplace` call.

Reply:
A copy is required. Without this, exceptions are thrown when accessing the object.

```
ctest_cell_formatted_as_date3 failed with:
xlnt::exception : bad attribute

ctest_timedelta failed with:
xlnt::exception : bad attribute

ctest_number_format failed with:
xlnt::exception : bad attribute

ctest_number_format_id failed with:
xlnt::exception : bad attribute

ftest_two_fills failed with:
xlnt::exception : bad attribute

stest_produce_simple_excel failed with:
```

Reply:
The exceptions are thrown as `next_custom_number_format_id` expects all `number_formats` having an id.

Calling `next_custom_number_format_id` before adding the new number format should work:
```
std::size_t new_id = next_custom_number_format_id();
iter = number_formats.emplace(number_formats.end(), new_number_format);
iter->id(new_id);
```